Java/syntax_test_java.java

```diff
@@ -398,6 +475,75 @@ void anonymousFunctions() {
 //                                  ^^^ storage.type.primitive - meta.annotation
 //                                      ^^^ variable.parameter.java
 //                                           ^^ storage.type.function.anonymous.java - meta.function.anonymous.parameters.java
+  }
+}
+
+class Generics {
+
+  List<String> field;
+//    ^^^^^^^^ meta.generic.java
+//    ^ punctuation.definition.generic.begin.java
+//     ^^^^^^ support.class.java
+//           ^ punctuation.definition.generic.end.java
+
+  List<java.net.URI> field;
+//    ^^^^^^^^^^^^^^ meta.generic.java
+//    ^ punctuation.definition.generic.begin.java
+//     ^^^^^^^^^^^^ meta.path.java
+//     ^^^^ support.type.package.java
+//         ^ punctuation.accessor.dot.java
+//          ^^^ support.type.package.java
+//             ^ punctuation.accessor.dot.java
+//              ^^^ support.class.java
+
+  void variableTypes() {
+    List<String> x;
+//      ^^^^^^^^ meta.generic.java
+//      ^ punctuation.definition.generic.begin.java
+//             ^ punctuation.definition.generic.end.java
+
+    List<java.lang.String> x;
+//      ^^^^^^^^^^^^^^^^^^ meta.generic.java
+//       ^^^^^^^^^^^^^^^^ meta.path.java
+//       ^^^^ support.type.package.java
+//           ^ punctuation.accessor.dot.java
+//            ^^^^ support.type.package.java
+//                ^ punctuation.accessor.dot.java
+//                 ^^^^^^ support.class.java
+//                       ^ punctuation.definition.generic.end.java
+
+    List<URI> x;
+//      ^^^^^ meta.generic.java
+//       ^^^ support.class.java
+
+    List<java.net.URI> x;
+//      ^^^^^^^^^^^^^^ meta.generic.java
+//       ^^^^^^^^^^^^ meta.path.java
+//                ^^^ support.class.java
+
+    List<int[]> x;
```

Comment:
Can you add a test for List<int> and make sure the primitive is marked as invalid.

Reply:
This construction is legal, you can use array as a type argument.

Reply:
Sorry I should have said, can you add a fest for List<int> x; as the int is invalid.

Reply:
Well, about primitives in generics. I actually opted out of using `invalid` scope in favor of just not highlighting them. This lonely lint feels out of place, especially in a language in which you won't go very far without a compiler. And it activates when you type an array, the behavior I find a bit unpleasant.

I'd really like to hear another opinions on this topic. I personally think that the syntax file should mostly care about the correct stuff. The compiler will handle incorrect one much better.

Reply:
> The compiler will handle incorrect one much better.

Yes. Don't use the `invalid` scope unless you're **absolutely, provable, 100% certain** that something is invalid. In practice this means you can scope unmatched parentheses/brackets/braces as invalid, but that's about it. Don't be too clever.

Reply:
Ah, I didn't think about the scopes applying white typing out the array.